pennylane/gradients/pulse_gradient.py

```diff
@@ -55,55 +56,65 @@ def _assert_has_jax(transform_name):
-            angle, the second entry the operations with negative Pauli rotation angle.
+            inner lists. The number of tuples is given by the number of shifted terms in the
+            parameter-shift rule of the generating Hamiltonian term ``ob``.
+        tensor_like: Coefficients of the parameter-shift rule of the provided generating Hamiltonian
+            term ``ob``.
     """
     t0, *_, t1 = op.t
-    if broadcast := qml.math.ndim(tau) > 0:
+    # If there are multiple values for tau, use broadcasting
+    if bcast := qml.math.ndim(tau) > 0:
+        # With broadcasting, create a sorted array of [t_0, *sorted(taus), t_1]
+        # Use this array for both, the pulse before and after the inserted operation.
+        # The way we slice the resulting tape results later on accomodates for the additional
+        # time points t_0 and t_1 in the array.
         tau = jnp.sort(tau)
         before_t = jnp.concatenate([jnp.array([t0]), tau, jnp.array([t1])])
         after_t = before_t.copy()
     else:
+        # Create a time interval from start to split and one from split to end
         before_t = jax.numpy.array([t0, tau])
         after_t = jax.numpy.array([tau, t1])
 
-    before_plus, before_minus = (
-        op(op.data, before_t, return_intermediate=broadcast, **op.odeint_kwargs) for _ in range(2)
-    )
-    after_plus, after_minus = (
-        op(
-            op.data,
-            after_t,
-            return_intermediate=broadcast,
-            complementary=broadcast,
-            **op.odeint_kwargs,
-        )
-        for _ in range(2)
-    )
+    if qml.pauli.is_pauli_word(ob):
+        prefactor = next(iter(qml.pauli.pauli_sentence(ob).values()))
+        word = qml.pauli.pauli_word_to_string(ob)
+        insert_ops = [qml.PauliRot(shift, word, ob.wires) for shift in [np.pi / 2, -np.pi / 2]]
+        coeffs = [prefactor, -prefactor]
+    else:
+        eigvals = qml.eigvals(ob)
+        coeffs, shifts = zip(*generate_shift_rule(eigvals_to_frequencies(tuple(eigvals))))
+        insert_ops = [qml.exp(qml.dot([-1j * shift], [ob])) for shift in shifts]
+
     # Create Pauli rotations to be inserted at tau
-    evolve_plus = qml.PauliRot(np.pi / 2, word, wires=word_wires)
-    evolve_minus = qml.PauliRot(-np.pi / 2, word, wires=word_wires)
-    # Construct gate sequences of split intervals and inserted Pauli rotations
-    ops = ([before_plus, evolve_plus, after_plus], [before_minus, evolve_minus, after_minus])
-    return ops
+    ode_kwargs = op.odeint_kwargs
+    # If we are broadcasting, make use of the `return_intermediate` and `complementary` features
+    ops = tuple(
+        [
+            op(op.data, before_t, return_intermediate=bcast, **ode_kwargs),
+            insert_op,
+            op(op.data, after_t, return_intermediate=bcast, complementary=bcast, **ode_kwargs),
+        ]
+        for insert_op in insert_ops
+    )
+    return ops, qml.math.array(coeffs)
```

Comment:
Here I would use jax directly for the coeffs, we had some trouble in the past with qml.math.array causing troubles with backprop.